### PR TITLE
Custom Vertex Buffer Serialization

### PR DESCRIPTION
**Pull request by carolhmj**
Closes https://github.com/BabylonJS/ThePirateCove/issues/547

With these changes, when serializing a mesh/scene, there are two new top level objects, `buffers` and `vertexBuffers`. Both are mappings from id (which was added to both as GUID) to serialized objects. On the geometry itself, instead of saving the buffer data, it saves the ID of the corresponding vertex buffer. This allows shared Buffers and VertexBuffers to be serialized/parsed only once.

Playground ID: #KSBUDT#12

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14749/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14749/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14749/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14749/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14749/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14749/merge#BCU1XR#0

**Review by Popov72 — COMMENTED** [review]
This seems generally correct to me, but I have a more general comment.

The current serialization of `Geometry` is not really correct, because if a `Buffer` is shared by several `VertexBuffer`, we will lose this information after a serialization/parsing process (since we only serialize the data of each vertex buffer).

Implementing it correctly would mean serializing a `Buffer` with all its data (basically, all the data needed by the constructor), so that we can rebuild it in a `parse` function.

So, I don't know if we should make an intermediate PR (this PR) to enable support for custom vertex buffer serialization, or if we should make a PR that will fix the problem outlined above (and add support for custom vertex buffer serialization in the meantime)...

I'd vote for the latter, but let's leave it to @sebavan.

**Comment by sebavan** [comment]
I would vote for the latter as well as this is not critical (not blocking anybody for now). 

Let s only address it once.

**Comment by Popov72** [comment]
> EDIT: Can a Buffer be shared between geometries? Because if so, then we have to serialize them in a level above the geometries...

Indeed, that's what I was about to comment... In glTF, for eg, it's possible to have big vertex buffers with mesh data (positions, normals, etc) pointing all to the same buffer but with different offsets.

So, the `Geometry` should only serialize a pointer to a buffer, and we should add a proper `serializing` / `parsing` method on the `Buffer` class.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Experimental playwright test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14749/merge/testResults/webgl2playwright/index.html

**Comment by sebavan** [comment]
@carolhmj will you finish this one or should we park it for later ?

**Comment by carolhmj** [comment]
@sebavan I can take a look at it whenever there is time 🙂

File: packages/dev/core/src/Buffers/buffer.ts [hunk 1]
```diff
@@ -3,6 +3,8 @@ import type { ThinEngine } from "../Engines/thinEngine";
 import { DataBuffer } from "./dataBuffer";
 import type { Mesh } from "../Meshes/mesh";
 import { Logger } from "../Misc/logger";
+import { RandomGUID } from "../Misc/guid";
+import { Tools } from "core/Misc/tools";
```
Comment (sebavan) [inline packages/dev/core/src/Buffers/buffer.ts:7]: Would be great to prevent this import for size footprint of thinEngine

**Comment by sebavan** [comment]
closing for now, let s reopen when we ll have more time

**Pull request closed without merging**